### ISSUE TITLE
How do I use my own car model?

### ISSUE THREAD

**Issue by n1ckfg** [Microsoft/AirSim#627]
Looking to change the specs of the default car model--is there anything addressing this in the documentation?

**Comment by mitchellspryn** [comment]
If you're just looking to change the physics and are OK with keeping the same visual model, the simplest thing would be to edit the parameters of the included uasset file for the SUV (in AirSim Content/VehicleAdv/SUV/)

If you have a completely custom mesh, then you can set it up like any other wheeled vehicle, but instead parent it to CarPawn. And then change the CarPawn constructor to point to the correct assets.

**Comment by n1ckfg** [comment]
Thanks! Yup, just looking to change the physics, don't care about the appearance of the geo for now. Have you got anything there that would affect the throttle command? (Is there a drivetrain model?)

**Comment by mitchellspryn** [comment]
For throttle, there are a few things you can try. 

First, open the blueprint SuvCarPawn, and navigate to the VehicleMovement component:

![image](https://user-images.githubusercontent.com/20287427/33153705-3ca49f60-cf98-11e7-9b8f-07cac1934e14.png)

There are a few relevant parameters you can play with:

- Max RPM: this controls the top speed of the car. 
- Torque Curve: this controls the acceleration. To up it, add a few more points and slide it up. 

There are drivetrain / transmission models, but those usually affect handling (i.e. turning) more than acceleration. 

Do keep in mind that there will need to be some manual tuning; I have found that copy-pasting actual data from real-world tests done on car components (e.g. torque curves from engines) usually doesn't give very good results.

**Comment by n1ckfg** [comment]
Awesome, I'll play with this!

**Comment by benjamin-lange** [comment]
Would you please provide a more detailed tutorial on how add a custom mesh vehicle to AirSim? Thank you!

**Comment by chioveri** [comment]
Hi, I would also be happy to have more infos on how to use a custom mesh car/vehicle. Thank you

**Comment by benjamin-lange** [comment]
@chioveri 

May be this works for you, I will try it out as soon as I find the time:
https://github.com/Microsoft/AirSim/blob/master/docs/settings.md
The documentation deals with all sorts of parameters that can be
set in the "/Documents/Airsim/settings.json file". Have a look at the 
"pawn paths" section.

**Comment by fernandodamasio** [comment]
@mitchellspryn 

About your comment:

> If you have a completely custom mesh, then you can set it up like any other wheeled vehicle, but instead parent it to CarPawn. And then change the CarPawn constructor to point to the correct assets.

Do we change the mesh in this code snippet here?

```
ACarPawn::ACarPawn()
{
    static ConstructorHelpers::FClassFinder<APIPCamera> pip_camera_class(TEXT("Blueprint'/AirSim/Blueprints/BP_PIPCamera'"));
    pip_camera_class_ = pip_camera_class.Succeeded() ? pip_camera_class.Class : nullptr;

    const auto& car_mesh_paths = AirSimSettings::singleton().pawn_paths["DefaultCar"];
```

Do you have any tips on how to do it? We need to change the mesh of the car, using another blueprint that I have.

Best Regards,
Fernando Damasio

**Comment by benjamin-lange** [comment]
@fernandodamasio
I did not look into your code problem.
But here is a skeleton settings.json for the custom model I use:
`
  "SettingsVersion": 1.2,
  "SimMode": "Car",
  "PawnPaths": {
	"DefaultCar": {"PawnBP":"Class'/Game/TruckTrailer/Truck/Blueprints/Truck.Truck_C'"}
	} 
`
You have to adjust the path in "DefaultCar" to your setup.
Please refer to the [documentation](https://github.com/Microsoft/AirSim/blob/master/docs/settings.md) for details. Look under "Other Settings" paragraph "PawnPaths".

**Comment by fernandodamasio** [comment]
> @fernandodamasio
> I did not look into your code problem.
> But here is a skeleton settings.json for the custom model I use:
> `"SettingsVersion": 1.2, "SimMode": "Car", "PawnPaths": { "DefaultCar": {"PawnBP":"Class'/Game/TruckTrailer/Truck/Blueprints/Truck.Truck_C'"} } `
> You have to adjust the path in "DefaultCar" to your setup.
> Please refer to the [documentation](https://github.com/Microsoft/AirSim/blob/master/docs/settings.md) for details. Look under "Other Settings" paragraph "PawnPaths".

Hi @benjamin-lange 

I tried to change only the mesh inside the cars' blueprint. It didn´t work. So, I guess I need another car to change it? I mean, I can't change just the mesh / appearance of the Airsim car.

Appreciate your help.

Best regards,
Fernando Damasio

**Comment by benjamin-lange** [comment]
> I tried to change only the mesh inside the cars' blueprint. It didn´t work.
I recommend leaving the default blueprint alone.
Just to clarify what you tried to do, you tried to change the mesh as it is shown in the following screenshot?

![image](https://user-images.githubusercontent.com/37362568/47415177-1cd91600-d773-11e8-84ed-c3b88e27869b.png)

If so then you might break the animations associated with the mesh which could cause more problems.
That is if you want to use the default vehicle in the future you have to fix these problems. 

> So, I guess I need another car to change it?
Yes, I recommend that. At least copy the original Blueprint located in project_name\Plugins\AirSim\Content\VehicleAdv\SUV and rename it before you experiment on it. ;-)
After you made your changes try out the pawn path method as stated above.

For your experiments with custom vehicle Blueprints in Airsim have a look into this [free asset ](https://www.unrealengine.com/marketplace/carsim-vehicle-dynamics) from the marktplace  or simply use the car model from the Blueprint vehicle template.

**Comment by fernandodamasio** [comment]
Hi @benjamin-lange 

Thank you very much for your quick response and also congrats for the Airsim project!

Regards,

**Comment by benjamin-lange** [comment]
@karanparikh91, have you tried using the "WASD" keys to move the vehicle?

**Comment by karanparikh91** [comment]
@benjamin-lange  thanks for your reply.

Turns out the car was not exported properly from Blender. 
My colleague found this blog post useful in fixing the problem
blog post: https://forums.unrealengine.com/development-discussion/content-creation/66676-help-blender-vehicle-to-ue4

I've got things working correctly now.

**Comment by thejeshk** [comment]
Hi @benjamin-lange 
![my_own_vehicle_skeleton](https://user-images.githubusercontent.com/33362523/75112373-80af6f80-5643-11ea-8e93-ceab4dcaae5a.PNG)


I have my own own project and my own vehicle model with the following skeleton [see image]. I have created a child blueprint class from AIRSIM carpawn and selected a mesh of my own vehicle. What else I should do in addition to that to use my vehicle ?  


If I try to play. Iam getting the following error. [image] 
![load_error](https://user-images.githubusercontent.com/33362523/75112386-9b81e400-5643-11ea-988b-05fffc3fd6f5.PNG) 

![my_own_vehicle_skeleton](https://user-images.githubusercontent.com/33362523/75112388-a3418880-5643-11ea-98ce-428563530c3f.PNG)


My json settings are follows. 

{
  "SettingsVersion": 1.2,
  "SimMode": "Car",
  "PawnPaths": {
    "DefaultCar": { "PawnBP": "Class'/Game/Mower/Blueprints/Mower_AIRSIM.Mower_AIRSIM'" }
  }
}

**Comment by thejeshk** [comment]
Hello @benjamin-lange, @karanparikh91 @mitchellspryn @fernandodamasio > 


I am having difficulties adding custom vehicle into my own project. 


Here's what I have done so far: Followed same steps as @karanparikh91 

**I have created my own project** 

1. Imported my new fbx asset for the car. **It has the same skeletal hierarchy but different bone names.** 
[Note: Is it important to have the same name bone names?] 
2. Has physics asset. Wheels are set to kinematic bodies. I'm using single hull convex collisions on all bodies for my mesh
3. Created a BP for my new skeletal mesh for the car. The BP I created has the CarPawn as its parent
4. Changed the Airsim settings.json file to include the pawn path to my new blueprint

When I hit play, I do see my new car flying from somewhere and dropping down (even though the player start is on the floor) But it doesn't seem to respond to any input when I hit the arrow keys. Am I missing a step or something else?

Thanks!
TK
